From 51606911d456bbcbdb85337a87dd35ce059bed4f Mon Sep 17 00:00:00 2001
From: Justin Waters <justin.waters@timesys.com>
Date: Wed, 9 Mar 2016 13:19:14 -0500
Subject: [PATCH 4/7] board: ge: bx50v3: Add support for ELO board

The ELO board uses the same Q7 module as the bx50v3 boards, but with
a different baseboard. However, there are only two main differences
that affect U-boot:

1) Dual channel LVDS display
2) Serial console

Signed-off-by: Justin Waters <justin.waters@timesys.com>
---
 arch/arm/cpu/armv7/mx6/Kconfig |  4 +++
 board/ge/bx50v3/Kconfig        |  2 +-
 board/ge/bx50v3/bx50v3.c       | 76 ++++++++++++++++++++++++++++++++++++++++++
 configs/ge_elo_defconfig       |  8 +++++
 configs/ge_elo_yocto_defconfig |  9 +++++
 include/configs/ge_bx50v3.h    |  8 +++++
 6 files changed, 106 insertions(+), 1 deletion(-)
 create mode 100644 configs/ge_elo_defconfig
 create mode 100644 configs/ge_elo_yocto_defconfig

diff --git a/arch/arm/cpu/armv7/mx6/Kconfig b/arch/arm/cpu/armv7/mx6/Kconfig
index 722e0d0..b93c661 100644
--- a/arch/arm/cpu/armv7/mx6/Kconfig
+++ b/arch/arm/cpu/armv7/mx6/Kconfig
@@ -60,6 +60,10 @@ config TARGET_CM_FX6
 config TARGET_EMBESTMX6BOARDS
 	bool "embestmx6boards"
 
+config TARGET_GE_ELO
+	bool "General Electric ELO"
+	select MX6Q
+
 config TARGET_GE_B450V3
 	bool "General Electric B450v3"
 	select MX6Q
diff --git a/board/ge/bx50v3/Kconfig b/board/ge/bx50v3/Kconfig
index 320b3b8..c88058b 100644
--- a/board/ge/bx50v3/Kconfig
+++ b/board/ge/bx50v3/Kconfig
@@ -1,4 +1,4 @@
-if TARGET_GE_B450V3 || TARGET_GE_B650V3 || TARGET_GE_B850V3
+if TARGET_GE_ELO || TARGET_GE_B450V3 || TARGET_GE_B650V3 || TARGET_GE_B850V3
 
 config IMX_CONFIG
 	default "board/ge/bx50v3/bx50v3.cfg"
diff --git a/board/ge/bx50v3/bx50v3.c b/board/ge/bx50v3/bx50v3.c
index cf2cd1a..8d3e47a 100644
--- a/board/ge/bx50v3/bx50v3.c
+++ b/board/ge/bx50v3/bx50v3.c
@@ -324,6 +324,8 @@ static iomux_v3_cfg_t const backlight_pads[] = {
 	/* Backlight enable for LVDS display */
 	MX6_PAD_GPIO_0__GPIO1_IO00 | MUX_PAD_CTRL(NO_PAD_CTRL),
 #define LVDS_BACKLIGHT_GP IMX_GPIO_NR(1, 0)
+	/* Set PWM to GPIO so backlight is full strength */
+	MX6_PAD_SD1_DAT3__GPIO1_IO21 | MUX_PAD_CTRL(NO_PAD_CTRL),
 };
 
 static void do_enable_hdmi(struct display_info_t const *dev)
@@ -344,6 +346,9 @@ static int detect_baseboard(struct display_info_t const *dev)
 	    IS_ENABLED(CONFIG_TARGET_GE_B650V3))
 		return 1;
 
+	if (IS_ENABLED(CONFIG_TARGET_GE_ELO))
+		return (0 == dev->addr);
+
 	return 0;
 }
 
@@ -369,6 +374,26 @@ struct display_info_t const displays[] = {{
 		.vmode          = FB_VMODE_NONINTERLACED
 } }, {
 	.bus	= -1,
+	.addr	= 0,
+	.pixfmt	= IPU_PIX_FMT_RGB24,
+	.detect	= detect_baseboard,
+	.enable	= NULL,
+	.mode	= {
+		.name           = "SHARP-LQ156M1LG21",
+		.refresh        = 60,
+		.xres           = 1920,
+		.yres           = 1080,
+		.pixclock       = 7851,
+		.left_margin    = 100,
+		.right_margin   = 40,
+		.upper_margin   = 30,
+		.lower_margin   = 3,
+		.hsync_len      = 10,
+		.vsync_len      = 2,
+		.sync           = FB_SYNC_EXT,
+		.vmode          = FB_VMODE_NONINTERLACED
+} }, {
+	.bus	= -1,
 	.addr	= 3,
 	.pixfmt	= IPU_PIX_FMT_RGB24,
 	.detect	= detect_hdmi,
@@ -559,6 +584,55 @@ static void setup_display_b850v3(void)
 		     IOMUXC_GPR3_HDMI_MUX_CTL_MASK);
 }
 
+static void setup_display_elo(void)
+{
+	struct mxc_ccm_reg *mxc_ccm = (struct mxc_ccm_reg *)CCM_BASE_ADDR;
+	struct iomuxc *iomux = (struct iomuxc *)IOMUXC_BASE_ADDR;
+
+	set_ldb_clock_source(1);
+
+	clrbits_le32(&mxc_ccm->cscmr2, MXC_CCM_CSCMR2_LDB_DI0_IPU_DIV);
+
+	imx_setup_hdmi();
+
+	/* Set LDB_DI0 as clock source for IPU_DI0 */
+	clrsetbits_le32(&mxc_ccm->chsccdr,
+			MXC_CCM_CHSCCDR_IPU1_DI0_CLK_SEL_MASK,
+			(CHSCCDR_CLK_SEL_LDB_DI0 <<
+			 MXC_CCM_CHSCCDR_IPU1_DI0_CLK_SEL_OFFSET));
+
+	/* Turn on IPU LDB DI0 clocks */
+	setbits_le32(&mxc_ccm->CCGR3, MXC_CCM_CCGR3_LDB_DI0_MASK);
+
+	enable_ipu_clock();
+
+	writel(IOMUXC_GPR2_BGREF_RRMODE_EXTERNAL_RES |
+	       IOMUXC_GPR2_DI1_VS_POLARITY_ACTIVE_HIGH |
+	       IOMUXC_GPR2_DI0_VS_POLARITY_ACTIVE_LOW |
+	       IOMUXC_GPR2_BIT_MAPPING_CH1_SPWG |
+	       IOMUXC_GPR2_DATA_WIDTH_CH1_24BIT |
+	       IOMUXC_GPR2_BIT_MAPPING_CH0_SPWG |
+	       IOMUXC_GPR2_DATA_WIDTH_CH0_24BIT |
+	       IOMUXC_GPR2_SPLIT_MODE_EN_MASK |
+	       IOMUXC_GPR2_LVDS_CH0_MODE_ENABLED_DI0 |
+	       IOMUXC_GPR2_LVDS_CH1_MODE_ENABLED_DI0,
+	       &iomux->gpr[2]);
+
+	clrsetbits_le32(&iomux->gpr[3],
+			IOMUXC_GPR3_LVDS0_MUX_CTL_MASK |
+			IOMUXC_GPR3_LVDS1_MUX_CTL_MASK |
+			IOMUXC_GPR3_HDMI_MUX_CTL_MASK,
+		       (IOMUXC_GPR3_MUX_SRC_IPU1_DI0 <<
+			IOMUXC_GPR3_LVDS0_MUX_CTL_OFFSET));
+
+	/* backlights off until needed */
+	imx_iomux_v3_setup_multiple_pads(backlight_pads,
+					 ARRAY_SIZE(backlight_pads));
+
+	gpio_direction_input(LVDS_POWER_GP);
+	gpio_direction_input(LVDS_BACKLIGHT_GP);
+}
+
 static void setup_display_bx50v3(void)
 {
 	struct mxc_ccm_reg *mxc_ccm = (struct mxc_ccm_reg *)CCM_BASE_ADDR;
@@ -642,6 +716,8 @@ int board_init(void)
 #if defined(CONFIG_VIDEO_IPUV3)
 	if (IS_ENABLED(CONFIG_TARGET_GE_B850V3))
 		setup_display_b850v3();
+	else if (IS_ENABLED(CONFIG_TARGET_GE_ELO))
+		setup_display_elo();
 	else
 		setup_display_bx50v3();
 #endif
diff --git a/configs/ge_elo_defconfig b/configs/ge_elo_defconfig
new file mode 100644
index 0000000..51835ce
--- /dev/null
+++ b/configs/ge_elo_defconfig
@@ -0,0 +1,8 @@
+CONFIG_ARM=y
+CONFIG_ARCH_MX6=y
+CONFIG_TARGET_GE_ELO=y
+# CONFIG_CMD_IMLS is not set
+# CONFIG_CMD_FLASH is not set
+CONFIG_CMD_GPIO=y
+CONFIG_SPI_FLASH=y
+CONFIG_SPI_FLASH_STMICRO=y
diff --git a/configs/ge_elo_yocto_defconfig b/configs/ge_elo_yocto_defconfig
new file mode 100644
index 0000000..1cd475c
--- /dev/null
+++ b/configs/ge_elo_yocto_defconfig
@@ -0,0 +1,9 @@
+CONFIG_ARM=y
+CONFIG_ARCH_MX6=y
+CONFIG_TARGET_GE_ELO=y
+# CONFIG_CMD_IMLS is not set
+# CONFIG_CMD_FLASH is not set
+CONFIG_CMD_GPIO=y
+CONFIG_SPI_FLASH=y
+CONFIG_SPI_FLASH_STMICRO=y
+CONFIG_YOCTO_IMAGE=y
diff --git a/include/configs/ge_bx50v3.h b/include/configs/ge_bx50v3.h
index 4b3c25e..8be4350 100644
--- a/include/configs/ge_bx50v3.h
+++ b/include/configs/ge_bx50v3.h
@@ -24,13 +24,21 @@
 #elif defined(CONFIG_TARGET_GE_B850V3)
 #define CONFIG_BOARD_NAME	"General Electric B850v3"
 #define CONFIG_DEFAULT_FDT_FILE	"mx6q-b850v3.dtb"
+#elif defined(CONFIG_TARGET_GE_ELO)
+#define CONFIG_BOARD_NAME	"General Electric ELO"
+#define CONFIG_DEFAULT_FDT_FILE	"imx6q-elo.dtb"
 #else
 #define CONFIG_BOARD_NAME	"General Electric BA16 Generic"
 #define CONFIG_DEFAULT_FDT_FILE	"imx6q-ba16.dtb"
 #endif
 
+#if defined(CONFIG_TARGET_GE_ELO)
+#define CONFIG_MXC_UART_BASE	UART4_BASE
+#define CONFIG_CONSOLE_DEV	"ttymxc3"
+#else
 #define CONFIG_MXC_UART_BASE	UART3_BASE
 #define CONFIG_CONSOLE_DEV	"ttymxc2"
+#endif
 
 #ifdef CONFIG_YOCTO_IMAGE
 #define CONFIG_BOOT_DIR	""
-- 
2.5.0

